In [20]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go
sns.set_style('darkgrid')
plt.rcParams['font.size'] = 14
plt.rcParams['figure.figsize'] = (9, 5)
plt.rcParams['figure.facecolor'] = '#00000000'

# Step 1: Select a real-world dataset


In [21]:
#lets first see what type of data do we have here
df = pd.read_csv('cwurData.csv')
df.head()

world_rank                            institution         country  \
0           1                     Harvard University             USA   
1           2  Massachusetts Institute of Technology             USA   
2           3                    Stanford University             USA   
3           4                University of Cambridge  United Kingdom   
4           5     California Institute of Technology             USA   

   national_rank  quality_of_education  alumni_employment  quality_of_faculty  \
0              1                     7                  9                   1   
1              2                     9                 17                   3   
2              3                    17                 11                   5   
3              1                    10                 24                   4   
4              4                     2                 29                   7   

   publications  influence  citations  broad_impact  patents   score  year  
0             1          1          1           NaN        5  100.00  2012  
1            12          4          4           NaN        1   91.67  2012  
2             4          2          2           NaN       15   89.50  2012  
3            16         16         11           NaN       50   86.17  2012  
4            37         22         22           NaN       18   85.21  2012

In [22]:
df2 = pd.read_csv("timesData.csv")
df2.head(10)

world_rank                        university_name                   country  \
0          1                     Harvard University  United States of America   
1          2     California Institute of Technology  United States of America   
2          3  Massachusetts Institute of Technology  United States of America   
3          4                    Stanford University  United States of America   
4          5                   Princeton University  United States of America   
5          6                University of Cambridge            United Kingdom   
6          6                   University of Oxford            United Kingdom   
7          8     University of California, Berkeley  United States of America   
8          9                Imperial College London            United Kingdom   
9         10                        Yale University  United States of America   

   teaching international  research  citations income total_score  \
0      99.7          72.4      98.7       98.8   34.5        96.1   
1      97.7          54.6      98.0       99.9   83.7        96.0   
2      97.8          82.3      91.4       99.9   87.5        95.6   
3      98.3          29.5      98.1       99.2   64.3        94.3   
4      90.9          70.3      95.4       99.9      -        94.2   
5      90.5          77.7      94.1       94.0   57.0        91.2   
6      88.2          77.2      93.9       95.1   73.5        91.2   
7      84.2          39.6      99.3       97.8      -        91.1   
8      89.2          90.0      94.5       88.3   92.9        90.6   
9      92.1          59.2      89.7       91.5      -        89.5   

  num_students  student_staff_ratio international_students female_male_ratio  \
0       20,152                  8.9                    25%               NaN   
1        2,243                  6.9                    27%           33 : 67   
2       11,074                  9.0                    33%           37 : 63   
3       15,596                  7.8                    22%           42 : 58   
4        7,929                  8.4                    27%           45 : 55   
5       18,812                 11.8                    34%           46 : 54   
6       19,919                 11.6                    34%           46 : 54   
7       36,186                 16.4                    15%           50 : 50   
8       15,060                 11.7                    51%           37 : 63   
9       11,751                  4.4                    20%           50 : 50   

   year  
0  2011  
1  2011  
2  2011  
3  2011  
4  2011  
5  2011  
6  2011  
7  2011  
8  2011  
9  2011

# Step 2: Perform data preparation & cleaning

In [23]:
og_copy = df.copy()

In [24]:
og_copy.isnull().sum()

world_rank                0
institution               0
country                   0
national_rank             0
quality_of_education      0
alumni_employment         0
quality_of_faculty        0
publications              0
influence                 0
citations                 0
broad_impact            200
patents                   0
score                     0
year                      0
dtype: int64

In [25]:
og_copy.broad_impact

0         NaN
1         NaN
2         NaN
3         NaN
4         NaN
        ...  
2195    969.0
2196    981.0
2197    975.0
2198    975.0
2199    981.0
Name: broad_impact, Length: 2200, dtype: float64

As we can see broad impact column has 200 null values. But we cannot drop these rows as it contain important information about the ranking of Universities. <br>
Lets see how distributed the data within this column is by using describe()

In [26]:
og_copy['broad_impact'].describe()

count    2000.000000
mean      496.699500
std       286.919755
min         1.000000
25%       250.500000
50%       496.000000
75%       741.000000
max      1000.000000
Name: broad_impact, dtype: float64

We can see here the deviation of data in this column is not abnormal. So we can fill these null values with the mean.

In [27]:
og_copy['broad_impact'].mean()

496.6995

In [28]:
og_copy['broad_impact'].fillna(value=496.6995,inplace=True)

In [29]:
og_copy.isnull().sum()

world_rank              0
institution             0
country                 0
national_rank           0
quality_of_education    0
alumni_employment       0
quality_of_faculty      0
publications            0
influence               0
citations               0
broad_impact            0
patents                 0
score                   0
year                    0
dtype: int64

In [30]:
og_copy.describe()

world_rank  national_rank  quality_of_education  alumni_employment  \
count  2200.000000    2200.000000           2200.000000        2200.000000   
mean    459.590909      40.278182            275.100455         357.116818   
std     304.320363      51.740870            121.935100         186.779252   
min       1.000000       1.000000              1.000000           1.000000   
25%     175.750000       6.000000            175.750000         175.750000   
50%     450.500000      21.000000            355.000000         450.500000   
75%     725.250000      49.000000            367.000000         478.000000   
max    1000.000000     229.000000            367.000000         567.000000   

       quality_of_faculty  publications    influence    citations  \
count         2200.000000   2200.000000  2200.000000  2200.000000   
mean           178.888182    459.908636   459.797727   413.417273   
std             64.050885    303.760352   303.331822   264.366549   
min              1.000000      1.000000     1.000000     1.000000   
25%            175.750000    175.750000   175.750000   161.000000   
50%            210.000000    450.500000   450.500000   406.000000   
75%            218.000000    725.000000   725.250000   645.000000   
max            218.000000   1000.000000   991.000000   812.000000   

       broad_impact      patents        score         year  
count   2200.000000  2200.000000  2200.000000  2200.000000  
mean     496.699500   433.346364    47.798395  2014.318182  
std      273.561032   273.996525     7.760806     0.762130  
min        1.000000     1.000000    43.360000  2012.000000  
25%      273.000000   170.750000    44.460000  2014.000000  
50%      496.699500   426.000000    45.100000  2014.000000  
75%      716.000000   714.250000    47.545000  2015.000000  
max     1000.000000   871.000000   100.000000  2015.000000

Next step is to check if each column is of appropriate data type or not.

In [31]:
og_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2200 entries, 0 to 2199
Data columns (total 14 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   world_rank            2200 non-null   int64  
 1   institution           2200 non-null   object 
 2   country               2200 non-null   object 
 3   national_rank         2200 non-null   int64  
 4   quality_of_education  2200 non-null   int64  
 5   alumni_employment     2200 non-null   int64  
 6   quality_of_faculty    2200 non-null   int64  
 7   publications          2200 non-null   int64  
 8   influence             2200 non-null   int64  
 9   citations             2200 non-null   int64  
 10  broad_impact          2200 non-null   float64
 11  patents               2200 non-null   int64  
 12  score                 2200 non-null   float64
 13  year                  2200 non-null   int64  
dtypes: float64(2), int64(10), object(2)
memory usage: 240.8+ KB


Columns YEAR HAS ITS data type wrong here. YEAR should be of DATE datatype, BUT SINCE DATE ONLY HAS YEAR VALUE, WE CAN ALSO CHANGE IT TO INT DATA TYPE.<br>
    lets convert the data type of this column

In [32]:
og_copy['year'] = og_copy['year'].astype('int')


In [33]:
og_copy

world_rank                            institution         country  \
0              1                     Harvard University             USA   
1              2  Massachusetts Institute of Technology             USA   
2              3                    Stanford University             USA   
3              4                University of Cambridge  United Kingdom   
4              5     California Institute of Technology             USA   
...          ...                                    ...             ...   
2195         996              University of the Algarve        Portugal   
2196         997                  Alexandria University           Egypt   
2197         998            Federal University of Ceará          Brazil   
2198         999                 University of A Coruña           Spain   
2199        1000        China Pharmaceutical University           China   

      national_rank  quality_of_education  alumni_employment  \
0                 1                     7                  9   
1                 2                     9                 17   
2                 3                    17                 11   
3                 1                    10                 24   
4                 4                     2                 29   
...             ...                   ...                ...   
2195              7                   367                567   
2196              4                   236                566   
2197             18                   367                549   
2198             40                   367                567   
2199             83                   367                567   

      quality_of_faculty  publications  influence  citations  broad_impact  \
0                      1             1          1          1      496.6995   
1                      3            12          4          4      496.6995   
2                      5             4          2          2      496.6995   
3                      4            16         16         11      496.6995   
4                      7            37         22         22      496.6995   
...                  ...           ...        ...        ...           ...   
2195                 218           926        845        812      969.0000   
2196                 218           997        908        645      981.0000   
2197                 218           830        823        812      975.0000   
2198                 218           886        974        812      975.0000   
2199                 218           861        991        812      981.0000   

      patents   score  year  
0           5  100.00  2012  
1           1   91.67  2012  
2          15   89.50  2012  
3          50   86.17  2012  
4          18   85.21  2012  
...       ...     ...   ...  
2195      816   44.03  2015  
2196      871   44.03  2015  
2197      824   44.03  2015  
2198      651   44.02  2015  
2199      547   44.02  2015  

[2200 rows x 14 columns]

lets join USA data from Timesdata and original data

In [34]:
og_copy.institution.drop_duplicates(inplace=True)

In [35]:
og_copy[og_copy['institution']=='Harvard University']

world_rank         institution country  national_rank  \
0              1  Harvard University     USA              1   
100            1  Harvard University     USA              1   
200            1  Harvard University     USA              1   
1200           1  Harvard University     USA              1   

      quality_of_education  alumni_employment  quality_of_faculty  \
0                        7                  9                   1   
100                      1                  1                   1   
200                      1                  1                   1   
1200                     1                  1                   1   

      publications  influence  citations  broad_impact  patents  score  year  
0                1          1          1      496.6995        5  100.0  2012  
100              1          1          1      496.6995        7  100.0  2013  
200              1          1          1        1.0000        2  100.0  2014  
1200             1          1          1        1.0000        3  100.0  2015

In [36]:
usa_og = og_copy[og_copy['country']=='USA']
usa_og.shape

(573, 14)

In [37]:
usa_df2 = og_times_copy[og_times_copy['country'] == 'USA']
usa_df2.shape

NameError: name 'og_times_copy' is not defined

In [ ]:
usa_df2.institution.drop_duplicates(inplace=True)

In [ ]:
usa_df2.shape

In [ ]:
final = usa_og.merge(usa_df2,on='institution',suffixes=['_left','right'],how = 'left',indicator = True)

In [ ]:
final = final[final['_merge']=='both']

In [ ]:
final.info()

In [ ]:
final

In [ ]:
final.shape

In [ ]:
final = final.groupby(final.index)[['world_rank_left', 'national_rank', 'quality_of_education',
                                    'alumni_employment', 'quality_of_faculty', 'publications', 'influence',
                                    'citations_left', 'broad_impact', 'patents', 'score', 'teaching',
                                    'research', 'citationsright']].agg('mean')

In [ ]:
final.columns

In [ ]:
final

Now we have cleaned the data. Got rid of null values, changed data types of columns where it was necessary.<br>
we can move on to the next step

# Step 3: Perform exploratory analysis & visualization



 Lets see which country has the most number of universties in top 100 world uni rankings
 

In [ ]:
top_50_data = og_copy.head(100)
top_50_unis_countrycount = top_50_data['country'].value_counts()
top_50_unis_countrycount = pd.DataFrame(top_50_unis_countrycount)
top_50_unis_countrycount.sort_values('country',ascending = True)

In [ ]:
fig = px.bar(top_50_unis_countrycount, y=top_50_unis_countrycount.index,
             x=top_50_unis_countrycount.country, orientation='h',color = top_50_unis_countrycount.index
             )
fig.update_layout(
    title='COUNT OF COUNTRIES IN TOP 100 WORLD UNIVERSITIES RANKINGS',
    xaxis_title='OCCURANCE',
    yaxis_title='COUNTRY'
)
fig.show()

We can clearly see USA dominating the TOP 100 WORLD UNIVERSITY RANKING chart with a presence of 58 times!!<br>
That means out of 100, 58 Unis listed here are from USA. <br>
That's why most of the students prefer USA over any other countries for MS,PHD's and higher studies.

## A small comparison between USA and INDIAN universities

In [ ]:
usa_ind_unis = og_copy[og_copy['country'].isin(['India','USA'])]
usa_ind_unis

Lets see how many times India and USA makes into top 1000

In [ ]:
top1000USAIND = usa_ind_unis.head(1000)
top1000USAIND_count = top1000USAIND['country'].value_counts()
top1000USAIND_count = pd.DataFrame(top1000USAIND_count)
top1000USAIND_count

In [ ]:
fig = go.Figure()
fig.add_trace(go.Bar(x=top1000USAIND_count.index, y=top1000USAIND_count.country,
                     text=top1000USAIND_count.country, width=0.2))

We can see INDIA only occurs 31 times while USA occurs 573 times, which makes us think about our education system.<br>
Lets go the next step and understand what is lacking in our education system!

Lets compare something that matters the most when it comes to universities and that is QUALITY OF EDUCATION

In [ ]:
QOE = usa_ind_unis.groupby('country')['quality_of_education'].sum()
QOE=pd.DataFrame(QOE)

In [ ]:
fig = go.Figure()
colors = ['Blue', 'red']
fig.add_trace(go.Bar(x=QOE.index, y=QOE.quality_of_education,
                     width=0.5, text=QOE.quality_of_education,
                     textposition='auto', marker=dict(color=colors)
                    ))
fig.update_layout(
    title={
        'text': 'QUALITY OF EDUCATION(INDIA VS USA)',
        'x': 0.5, 
        'xanchor': 'center' 
    },
    xaxis_title='COUNTRIES',
    yaxis_title='QUALITY OF EDUCATION',

)

Here, there's humongous difference between quality of education in INDIA vs quality of education in USA. <br>
We can say USA's quality of Education is more than 10x better than India's. <br>

Lets compare how we do in  quality of faculty, publications, influence, citations, broad impact, patents, score compared to USA

In [ ]:
metrics = usa_ind_unis.groupby('country')[['quality_of_faculty',
                                         'publications', 'influence', 'citations', 'broad_impact', 'patents',
                                         'score']].sum()

In [ ]:
pd.DataFrame(metrics)

We can already see USA outperforming INDIA in each aspect, but lets understand this with some plots.

In [19]:
# Create a grouped bar chart
fig = go.Figure()
countries = metrics.index
variables = list(metrics.columns)

for variable in variables:
    fig.add_trace(go.Bar(
        x=countries,
        y=metrics[variable],
        name=variable,
        text=metrics[variable],
        textposition='outside',
    ))

# Customize the layout
fig.update_layout(
    title={'text': 'Comparison of Variables between Countries',
           'x': 0.5, 'xanchor': 'center'},
    xaxis_title='Country',
    yaxis_title='Value',
    barmode='group',
    width = 1000,
    height = 600,
    template = 'plotly_dark'
)

# Display the plot
fig.show()

NameError: name 'metrics' is not defined

This chart it sums it all pretty well.<br>
This chart should answer each and every question related to why INDIA lacks behind compared to USA in WORLD UNIVERSITY RANKINGS.<br>
This is the sum of each column i.e for  quality of faculty, publications, influence, citations, broad impact, patents, score of both the nations.<br>
We can consider these values as points, so each bar represents total points for that respective column.

## Lets see which University from which Country has the most number of Patents

In [ ]:
og_copy['patents'].describe()

In [ ]:
country_patents = og_copy.groupby('country')['patents'].sum()
country_patents = pd.DataFrame(country_patents)
country_patents  = country_patents.sort_values('patents',ascending=True)

In [ ]:
import random

In [ ]:
fig = go.Figure()
fig.add_trace(go.Bar(x=country_patents.patents,
                     y=country_patents.index, orientation='h',
                     width=0.6, text=country_patents.patents, textfont=dict(family='Arial', size=10),
                     marker=dict(color=[f'rgb({random.randint(0, 255)},{random.randint(0, 255)},{random.randint(0, 255)})' for _ in range(
                         len(country_patents.index))]),
                     textposition='outside'))
fig.update_layout(
    height=1000,
    template='plotly',
    title={'text': 'PATENTS PER COUNTRY', 'x': 0.5, 'xanchor': 'center'},
    xaxis_title='NUMBER OF PATENTS',
    yaxis_title='COUNTRIES'

)

We can see clearly, USA again dominating the most number of patents for a country category.

# Step 4: Ask & answer questions about the data
|

# Step 5: Summarise your inferences & write a conclusion
